Various server side incompatibility fixes

WilliamMcCumstie opened the pull request:
This PR is to track various changes required to the server so it behaves nicely.

Fix 1: Allow `jobs` to be created from a string version of `min_nodes`. Previously it was returning the following error:
```
bin/scheduler: The API returned a 403 error status and this content:
  {
    "errors": [
      {
        "id": "2d7c77c5-8b21-41c6-94f3-d9f1849686e8",
        "title": "Forbidden Error",
        "detail": "Client-generated ID not provided",
        "status": "403"
      }
    ]
  }
```
This is a red herring, there is no  "client generated ID" issue. Instead an `ArgumentError` has been raised as `min_nodes` was supplies as a `string` instead of an `integer`. The client now allows `k` and `m` suffixes to `min_nodes` and ipso facto submits everything as a string. Two  changes have been made in regard to this:
* The server now converts the `k` and `m` so that `min_nodes` ends up as an `integer` https://github.com/openflighthpc/flight-scheduler-controller/commit/e3dd8a695290f89c4c0befec3f7b917960c3d2ad
* The server now allows `min_nodes` to be a string version of an integer when the suffix is omitted: https://github.com/openflighthpc/flight-scheduler-controller/commit/4d684feeae7ac162c7748d9c3df1f3242f368d9d

Fix 2: Serialise Job against Partition instead of Scheduler
This has been done in various commits and is reasonable straight forward. The `Scheduler` model has been removed entirely and instead each `Job` is directly linked to a `Partition`.

Fix 3: Update the find method for jobs
The following error was previously returned:
```
bin/scheduler: The API returned a 500 error status and this content:
  {
    "errors": [
      {
        "id": "b952ecc1-58b1-4f72-8fb0-157fc41c74df",
        "title": "Name Error",
        "detail": "uninitialized constant App::PARTITIONS\nDid you mean?  Partition",
        "status": "500"
      }
    ]
  }
```
Also allow jobs to be cancelled using the new `FlightScheduler`: https://github.com/openflighthpc/flight-scheduler-controller/pull/2/commits/41951b1ba702e79d081861713746958f18f8827b

WilliamMcCumstie committed:
Allow min_nodes to be suffixed with k or m

This specifies a multiple of the number of nodes

Also removed the redundant rioSchedular swagger definitions

WilliamMcCumstie committed:
Handle when min_nodes is a stringified integer

This is very common use case from the client due to the k/m prefix.

However the partitions object fails to queue the job due to a type error

WilliamMcCumstie committed:
Remove schedular from JobSerializer

It no longer exists

WilliamMcCumstie committed:
Also remove the SchedularSerializer entirely

It also does not exist anymore

WilliamMcCumstie committed:
Serialize the Partition against a Job

This allows the client to access which partition each job is on

WilliamMcCumstie committed:
Move the index route out of the helper block

Technically this doesn't make a difference, but it has been bugging me

WilliamMcCumstie committed:
Document that script and min-nodes is required

This change broke the client

WilliamMcCumstie committed:
Allow jobs to be indexed and removed (again)

They where once broke and now they are fix

benarmston commented on app/models/job.rb:
```diff
@@ -36,6 +36,19 @@ class Job
   attr_accessor :state
 
+  # Handle the k and m suffix
+  attr_reader :min_nodes
```
The `attr_accessor :min_nodes` line above this one should be removed.

WilliamMcCumstie replied:
https://github.com/openflighthpc/flight-scheduler-controller/pull/2/commits/6ff71860dc4ceb1860e7156c3ab83b952b66ab5f

benarmston commented on app/models/job.rb:
```diff
@@ -36,6 +36,19 @@ class Job
+    elsif /\A\d+m\Z/.match(str)
+      str.sub('m', '').to_i * 1048576
+    else
+      num
```
If we change this to `num.to_i` I think we can remove the `to_i` from the various `node.min_nodes.to_i` calls elsewhere.  That way once the value has been set on the Job we're always dealing with a number afterwards.

WilliamMcCumstie replied:
I was contemplating that, but I don't know how nicely it will handle `min_num: "some-bogus-string"`. It will essentially defeat the purpose of the validation

benarmston replied:
OK. I see what you mean.  Let's leave it as it for now.  I may have a think on this some more.

WilliamMcCumstie replied:
Just realised we can use `regex` (like the other regex) :facepalm: 

https://github.com/openflighthpc/flight-scheduler-controller/pull/2/commits/3c129c69432c782950ac1a2c3ef7d31c85e220b4
https://github.com/openflighthpc/flight-scheduler-controller/pull/2/commits/bf4fd9b6fe2dc8705dc93ce4dfaadda56b2793d1

benarmston reviewed (changes requested):
A couple of minor issues.  Otherwise looks good.

WilliamMcCumstie committed:
Remove the redundant min_nodes attr_accessor

WilliamMcCumstie committed:
Pseudo require the client to send an argument array

These will be provided to the script when they are ran

It is flagged as required as in practice it will always be provided
by the client, even if empty. This is despite it not being validated
ATM

WilliamMcCumstie committed:
Type cast whole number strings to integers

This removes the various to_i method calls whilst preserving
the error message if `min_nodes` is a "bogus-string"

WilliamMcCumstie committed:
Ensure the entire `min_nodes` string is matched

benarmston commented on app/models/job.rb:
```diff
@@ -31,10 +31,27 @@ class Job
   attr_accessor :partition
   attr_accessor :script
   attr_accessor :state
+  attr_accessor :args
```
We already have `attr_accessor arguments` set for `Job`.  I prefer `arguments`.

WilliamMcCumstie replied:
Opps, completely missed it

https://github.com/openflighthpc/flight-scheduler-controller/pull/2/commits/5e3568c41a14271540d5d886a61f6cb4ff21a7d3

WilliamMcCumstie committed:
Use the pre-existing arguments key instead of args

benarmston commented:
LGTM.
